### PR TITLE
Implement IntegerCache semantics compliant with JLS

### PR DESCRIPTION
**Pull request by fhowar**
IntegerCache leads to unintended concretization

Addresses the issue reported in [https://github.com/tudo-aqua/gdart/issues/5](https://github.com/tudo-aqua/gdart/issues/5)

This PR implements the symbolic semantics described in the JLS for an IntegerCache of size 256.

### CODE REVIEW

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java [hunk 1]
```diff
@@ -804,6 +805,54 @@ public void takeBranchPrimitive2(VirtualFrame frame, BytecodeNode bcn, int bci,
...
+        if ((a1 == null) && (a2 == null)) {
+            return;
+        }
+
+        Expression expr = null;
+        Meta meta = bcn.getMeta();
+
+        // special case: cached Integer
+        if (meta.java_lang_Integer.equals(c1.getKlass()) &&
+                meta.java_lang_Integer.equals(c1.getKlass())) {
```
Comment (mmuesly) [inline espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: I think one of these should be 'c2'?

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java [hunk 2]
```diff
@@ -804,6 +805,54 @@ public void takeBranchPrimitive2(VirtualFrame frame, BytecodeNode bcn, int bci,
...
+
+            Expression e1 = c1 == null ? null : Annotations.annotation(
+                    AnnotatedVM.getFieldAnnotation(c1, meta.java_lang_Integer_value), config.getConcolicIdx());
+            Expression e2 = c2 == null ? null : Annotations.annotation(
+                    AnnotatedVM.getFieldAnnotation(c2, meta.java_lang_Integer_value), config.getConcolicIdx());
+
+            if (e1 != null || e2 != null) {
+
+                int int1 = meta.java_lang_Integer_value.getInt(c1);
+                int int2 = meta.java_lang_Integer_value.getInt(c1);
```
Comment (mmuesly) [inline espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: 'c2'?

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java [hunk 3]
```diff
@@ -804,6 +805,54 @@ public void takeBranchPrimitive2(VirtualFrame frame, BytecodeNode bcn, int bci,
...
+
+                e1 = e1 == null ? Expression.fromConstant(Types.INT, int1) : e1;
+                e2 = e2 == null ? Expression.fromConstant(Types.INT, int1) : e2;
```
Comment (mmuesly) [inline espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: 'int2'?

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java [hunk 4]
```diff
@@ -804,6 +805,54 @@ public void takeBranchPrimitive2(VirtualFrame frame, BytecodeNode bcn, int bci,
...
+
+                expr = new ComplexExpression(BAND,
+                        new ComplexExpression(BVEQ, e1, e2),
+                        new ComplexExpression(BVLE, Expression.fromConstant(Types.INT, 127), e1),
+                        new ComplexExpression(BVLE, e1, Expression.fromConstant(Types.INT, 127)));
```
Comment (mmuesly) [inline espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: Should there be one 'e2' in these BVLE expressions? Otherwise, this can only be true if, e1 = 127. Or one of the constants probably should be -128?

**Review by mmuesly — CHANGES_REQUESTED** [review]

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java (hunk 1 above)
Comment (fhowar) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: fixed

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java (hunk 2 above)
Comment (fhowar) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: fixed

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java (hunk 3 above)
Comment (fhowar) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: fixed

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java (hunk 4 above)
Comment (fhowar) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java]: fixed and extended

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java [hunk 5]
```diff
@@ -804,6 +805,51 @@ public void takeBranchPrimitive2(VirtualFrame frame, BytecodeNode bcn, int bci,
...
+
+                expr = new ComplexExpression(BAND,
+                        new ComplexExpression(BVEQ, e1, e2),
+                        new ComplexExpression(BVLE, Expression.fromConstant(Types.INT, -128), e1),
+                        new ComplexExpression(BVLE, e1, Expression.fromConstant(Types.INT, 127)),
+                        // remaining not strictly necessary?
+                        new ComplexExpression(BVLE, Expression.fromConstant(Types.INT, -128), e2),
+                        new ComplexExpression(BVLE, e2, Expression.fromConstant(Types.INT, 127)));
+
+                if (!(int1 == int2 && -128 <= int1 && int1 <= 127)) {
```
Comment (mmuesly) [inline espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java:847]: This check is only implemented on int1 while the expression affects int2. Is this mismatch intended?

Comment (fhowar) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java:847]: if int1 == int2 then what is tested afterwards holds for int1 and int2. I was more verbose in the symbolic constraint

**Comment by m-carrasco** [comment]
Thanks for writing this PR for the reported issue! I can confirm that it solves the reported issue.

Here is the test case to reproduce the concretization:

```
import tools.aqua.concolic.Verifier;

public class Repro {
    public static void main(String[] args) {
        int symbolicBV = Verifier.nondetInt();
        Integer boxed = Integer.valueOf(symbolicBV);

        if (boxed == 10){
            System.out.println("Equal to ten!");
        } else {
            System.out.println("They are not equal!");
        }

    }
}
```

Prior to this PR, the generated path conditions were:

```
; This is checking -128 <= symbolic bv <= 127
[DECLARE] (declare-fun __int_0 () (_ BitVec 32))
[WITNESS] Repro.java : LRepro;.main([Ljava/lang/String;)V : 8 : 0
[DECISION] (assert (bvsle #xffffff80 __int_0)) // branchCount=2, branchId=0
[DECISION] (assert (bvsge #x0000007f __int_0)) // branchCount=2, branchId=0
```

Using the latest changes in this PR (9e39a53012a60ba0e4965fda8fa93d016e515d05), the generated path conditions were:

```
[DECLARE] (declare-fun __int_0 () (_ BitVec 32))
[WITNESS] Repro.java : LRepro;.main([Ljava/lang/String;)V : 8 : 0
[DECISION] (assert (!= #x0000000a __int_0)) // branchCount=2, branchId=1
```

The above conditions look correct to me and solve the reported issue.

----------------------

I find the following weird, and I cannot fully understand it. Could you let me know what you think about it?

Consider the path conditions generated before this PR; they were just checking that `-128 <= symbolic bv <= 127`. However, I would also expect to find path conditions for accessing the cache. It must have traversed a data structure using the symbolic value to find the corresponding constant cached value. In other words, I would expect the path conditions to reflect the cache lookup access even though `boxed` is mapped to a constant symbolic expression (the cache result). 

One reason to explain this is that the cache lookup somewhat concretizes expressions, too. One simple case could be native calls, but I'm not sure the cache lookup triggers that.

**Comment by m-carrasco** [comment]
I also tested the following case (using zero as seed), which looks correct --- using this PR:

```
public class Repro1 {
    public static void main(String[] args) {
        int symbolicBV = Verifier.nondetInt();
        Integer boxedOne = Integer.valueOf(symbolicBV);
        Integer boxedTwo = GetZero();

        if (boxedOne == boxedTwo){
            System.out.println("Both equal!");
        } else {
            System.out.println("Not equal");
        }

        System.exit(0);
    }

    public static Integer GetZero(){
        return Integer.valueOf(0);
    }
}
```

The output was:

```
Both equal!
======================== END PATH [BEGIN].
[DECLARE] (declare-fun __int_0 () (_ BitVec 32))
[WITNESS] Repro1.java : LRepro1;.main([Ljava/lang/String;)V : 8 : 0
[DECISION] (assert (and (= #x00000000 __int_0) (bvsle #xffffff80 #x00000000) (bvsle #x00000000 #x0000007f) (bvsle #xffffff80 __int_0) (bvsle __int_0 #x0000007f))) // branchCount=2, branchId=0
======================== END PATH [END].
```

This is a minor thing, but we have concrete comparisons in the constraints, which could be redundant.

File: espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java [hunk 6]
```diff
@@ -804,6 +805,51 @@ public void takeBranchPrimitive2(VirtualFrame frame, BytecodeNode bcn, int bci,
...
+
+        // special case: cached Integer
+        if (meta.java_lang_Integer.equals(c1.getKlass()) &&
+                meta.java_lang_Integer.equals(c2.getKlass())) {
+
+            Expression e1 = c1 == null ? null : Annotations.annotation(
+                    AnnotatedVM.getFieldAnnotation(c1, meta.java_lang_Integer_value), config.getConcolicIdx());
+            Expression e2 = c2 == null ? null : Annotations.annotation(
+                    AnnotatedVM.getFieldAnnotation(c2, meta.java_lang_Integer_value), config.getConcolicIdx());
+
```
Comment (m-carrasco) [inline espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java:830]: Could we avoid computing a resulting symbolic expression if both operands' expressions are symbolic constant values or don't have one?

Comment (mmuesly) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java:830]: Hi, I don't get this question. If they are symbolic constants, we still need the expression to model effects happening in other part of the program. If they are both constants without symbolic annotation, the if in line 810 ensures, that we do not do any further symbolic processing.

Could you make an example what exactly you are looking for?

Comment (m-carrasco) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java:830]: Thanks for your response! I’m really interested in learning more about SPouT, and I appreciate your insights. I realize now that I overlooked the check at line 810 for cases where neither operand has symbolic annotations— thanks for pointing that out.

Regarding the scenario where both operands have symbolic constants as annotations, I didn’t mean to imply that creating a path constraint for such a case is a bug. Rather, I was just wondering if it’s strictly necessary.

For example, pushing a path condition like 5 > 4 doesn’t seem to enable exploration of new paths compared to either not pushing anything or just pushing true. Similarly, associating the result’s stack slot with the symbolic expression 5 > 4 wouldn’t be advantageous. That was what I thought while reading the code. I apologize if I misunderstood how the code behaves!

Comment (mmuesly) [reply espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java:830]: In case, c1 or c2 has a symbolic annotation, the constraint would look more likely like `int_1 > 4`, if the other is concrete, or something like `int_1 > int_2` if booth of them are symbolic. We do not have path constraints without any symbolic variables in SPouT and in case there are any (which I would consider a bug), we would reject them in the DSE component building the constraint tree.

**Comment by mmuesly** [comment]
I integrated these changes together with some more changes for SV-COMP 26 into the spout-22.2-jdk17 branch.
